packages/dds/tree/docs/wip/schema-evolution.md

```diff
@@ -0,0 +1,583 @@
+
+### Migration
+
+Many developers who change their schema will need to think carefully about how they migrate their application from the old schema to the new schema.
+It requires thoughtful planning in a production scenario.
+
+An application has its schema encoded in multiple places.
+
+* The **stored schema** is the schema that the tree in the **document** - the persisted state and source of truth for the application's data, shared by all clients - conforms to.
+  It is persisted as part of the document.
+  All clients reading the document know that the tree stored in that document matches this schema.
+* The **view schema** is the schema that a client expects when reading and writing tree.
+  It is part of the application source code loaded by a client.
+
+When an application is first written, these two schema are equivalent.
+In some environments, the developer can ensure that they remain equivalent as new versions are released.
+For example, consider an application where all the documents are stored on a central server.
+To release a new version, a developer takes the following steps to upgrade from schema A to schema B:
+
+1. At first, the documents are using stored schema A and the application is written against stored schema A.
+2. The developer temporarily takes the service offline, suspending application availability, and then upgrades all documents on the server to use stored schema B before re-enabling the service.
+   Thereafter, any clients attempting to connect to the service will be prompted to download the new version of the application client first, which is written against view schema B.
+3. The migration is complete.
+
+That makes a clean and relatively simple transition.
+The diagram below visualizes the clean break between the old and new schema that happens at step 2:
+
+```mermaid
+gantt
+dateFormat D
+axisFormat Step %e
+tickInterval 1day
+Maintain view schema A in application source code : 1, 1d
+Documents contain stored schema A                 : 1, 1d
+Maintain view schema B in application source code : 2, 1d
+Documents contain stored schema B                 : 2, 1d
+```
+
+Unfortunately, many applications operate under additional constraints.
+For example, what if the developer does not have control over all the documents?
+Perhaps the documents can be stored locally on a user's computer, or can be hosted by another service that is independent of the developer.
+What if the developer does not have control over when a client updates their application?
+Even "live service" applications, like web apps, might allow users to continue using an older/cached version of an application client long after a deployment of a new version.
+In these scenarios, when does the document get upgraded?
+When does the client move to the new version of the application?
+The developer cannot know, and therefore cannot guarantee that the stored schema and view schema of their application will be the same for all clients and documents.
```

Comment:
I think you are missing the main reasons:
- it's very risky (if this process goes wrong, data is likely lost). There is no way to adhere to safe velocity principles - it's all or nothing.
- in most environments, it's just too expensive (it would cost millions if not billions of dollars to do that at scale of SharePoint, for example convert all Word documents).

Many of the points you raised could be somewhat addressed. For example, application (of specific version) can be kicked out from collab session when upgrade for a document happens. As long as there are "before" and "after" versions of the app available, and "after" version was available for long enough time that it reached all users.